### PR TITLE
Create proposal validation

### PR DESCRIPTION
**Pull request by dev-matan-tsuberi**
resolve #18 

Question: is there a better place to put validation logic?

### CODE REVIEW

**Comment by dev-matan-tsuberi** [comment]
[Validation of create proposal form](https://trello.com/c/oLQ1C6Wg/73-validation-of-create-proposal-form)

File: src/actions/arcActions.ts [hunk 1]
```diff
@@ -217,7 +217,7 @@ export async function getDAOData(avatarAddress : string, currentAccountAddress :
         await new Promise((resolve) => {
           eventFetcher.get((err, events) => {
             if (typeof err === 'undefined' && events.length > 0) {
-              genesisProposal.reputationWhenExecuted = Number(web3.fromWei(events[0].args._totalReputation, "ether"));
+              genesisProposal.reputationWhenExecuted = Number(web3.fromWei((events[0].args as any)._totalReputation, "ether"));
```
Comment (dev-matan-tsuberi) [inline src/actions/arcActions.ts]: https://github.com/daostack/arc.js/issues/129

Comment (tibetsprague) [reply src/actions/arcActions.ts]: this has been fixed on master already, please update to the latest master

File: src/actions/arcActions.ts [hunk 2]
```diff
@@ -452,8 +452,11 @@ export function createProposal(daoAvatarAddress : string, title : string, descri
     dispatch({ type: arcConstants.ARC_CREATE_PROPOSAL_PENDING, payload: null });
     try {
       const web3 : Web3 = Arc.Utils.getWeb3();
-      const ethAccountAddress : string = getState().web3.ethAccountAddress;
 
+      if(!web3.isAddress(beneficiary)) throw new Error(`Invalid address`);
```
Comment (tibetsprague) [inline src/actions/arcActions.ts]: how does this error get caught and handled on the front end?

Comment (dev-matan-tsuberi) [reply src/actions/arcActions.ts]: It's caught and dispatched as a `ARC_CREATE_PROPOSAL_REJECTED`.
How would you like to handle it?

File: src/actions/arcActions.ts (hunk 1 above)
Comment (dev-matan-tsuberi) [reply src/actions/arcActions.ts]: updated: https://github.com/daostack/alchemy/pull/18/commits/767e38aac36777be3736bc59365ba65c66acde49

File: package.json [hunk 3]
```diff
@@ -24,7 +24,7 @@
     "auto-start-ganache": "cross-env ETH_ENV=ganache run-with-ganache --ganache-cmd ganache-cli \"npm run migrate-ganache && npm run start-ganache\""
   },
   "dependencies": {
-    "@daostack/arc.js": "0.0.0-alpha.32",
+    "@daostack/arc.js": "^0.0.0-alpha.33",
```
Comment (tibetsprague) [inline package.json]: im seeing bugs with version 33, so I dont want to upgrade right now

File: src/actions/arcActions.ts (hunk 2 above)
Comment (tibetsprague) [reply src/actions/arcActions.ts]: ok, that works, i mean more that an error should be displayed in the UI

File: package.json (hunk 3 above)
Comment (tibetsprague) [reply package.json]: actually this is ok, i got it to work

File: src/actions/arcActions.ts (hunk 2 above)
Comment (dev-matan-tsuberi) [reply src/actions/arcActions.ts]: update: https://github.com/daostack/alchemy/pull/18/commits/005286a25b051811ed007e726f8871c8e309c778

Moved validation to form itself.

File: src/actions/arcActions.ts [hunk 4]
```diff
@@ -664,7 +666,6 @@ export function stakeProposal(daoAvatarAddress: string, proposalId: string, pred
       if(amount < minimumStakingFee) throw new Error(`Staked less than the minimum: ${minimumStakingFee}!`);
       if(amount > balance) throw new Error(`Staked more than than the balance: ${balance}!`);
 
-      await stakingToken.approve(votingMachineInstance.address, amount);
```
Comment (tibetsprague) [inline src/actions/arcActions.ts]: why did you remove this? wasn't this necessary?

File: src/components/CreateProposal/CreateProposal.scss [hunk 5]
```diff
@@ -64,8 +64,16 @@
   		border-radius: 3px;
   		font-size: $subheader;
   		padding: 8px;
-      font-family: $body-font;
-  	}
+			font-family: $body-font;
```
Comment (tibetsprague) [inline src/components/CreateProposal/CreateProposal.scss]: man the whitespace in this doc is wack, ideally you should stick with the current system though

File: src/components/CreateProposal/CreateProposal.scss [hunk 6]
```diff
@@ -64,8 +64,16 @@
...
+
+			&.error{
+				border: 1px solid red;
+			}
+		}
+
+		.errorMessage{
```
Comment (tibetsprague) [inline src/components/CreateProposal/CreateProposal.scss]: please put spaces before open braces

File: src/components/CreateProposal/CreateProposalContainer.tsx [hunk 7]
```diff
@@ -83,8 +101,11 @@ class CreateProposalContainer extends React.Component<IProps, IState> {
 
   handleSubmit = (event : any) => {
     event.preventDefault();
-    this.setState({ buttonEnabled: false });
-    this.props.createProposal(this.state.avatarAddress, this.state.title, this.state.description, this.state.nativeTokenReward, this.state.reputationReward, this.state.beneficiary);
+
+    if(!Object.keys(this.state.errors).length){
```
Comment (tibetsprague) [inline src/components/CreateProposal/CreateProposalContainer.tsx]: more whitespace in general please, like between if and ( and between ) and {

File: src/components/CreateProposal/CreateProposalContainer.tsx [hunk 8]
```diff
@@ -159,14 +180,22 @@ class CreateProposalContainer extends React.Component<IProps, IState> {
           </label>
           <input
             id='beneficiaryInput'
+            className={this.state.errors.beneficiary ? css.error : null}
             maxLength={42}
             onChange={this.handleChange}
+            onBlur={e => this.validate()}
```
Comment (tibetsprague) [inline src/components/CreateProposal/CreateProposalContainer.tsx]: seems like you could build this into the onChange handler

File: src/actions/arcActions.ts (hunk 4 above)
Comment (dev-matan-tsuberi) [reply src/actions/arcActions.ts]: Arc.js takes care of this now.

File: src/components/CreateProposal/CreateProposalContainer.tsx (hunk 8 above)
Comment (dev-matan-tsuberi) [reply src/components/CreateProposal/CreateProposalContainer.tsx]: I think it's annoying to have it shout "invalid address!" at you while you type, don't you think?

File: src/components/CreateProposal/CreateProposalContainer.tsx (hunk 7 above)
Comment (dev-matan-tsuberi) [reply src/components/CreateProposal/CreateProposalContainer.tsx]: Should we add a linter and settle on specific linting rules?

File: src/components/CreateProposal/CreateProposal.scss (hunk 5 above)
Comment (dev-matan-tsuberi) [reply src/components/CreateProposal/CreateProposal.scss]: Seems like my editor messed it up somehow...

File: src/components/CreateProposal/CreateProposalContainer.tsx (hunk 7 above)
Comment (tibetsprague) [reply src/components/CreateProposal/CreateProposalContainer.tsx]: yes we should, there's a ticket for that. but not this second

File: src/components/CreateProposal/CreateProposalContainer.tsx (hunk 8 above)
Comment (tibetsprague) [reply src/components/CreateProposal/CreateProposalContainer.tsx]: hmm, yes thats true, I forgot that onChange fires all the time
